### PR TITLE
Allow specification of matplotlib figure for circuit drawing

### PR DESCRIPTION
**Pull request by eendebakpt**
**Context:**

The current implementation of `qmpl.draw_mpl` always creates a new figure window. In this PR we allow the figure window to be specified. For example

```
import pennylane
import pennylane as qml
import matplotlib.pyplot as plt

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit(params):
    qml.RY(params[0], wires=0)
    qml.RY(params[1], wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(1))

params = [0.1, 0.3]

qml.draw_mpl(circuit, decimals=2, fig=plt.figure(10))(params) # set figure window number 10
# ...
qml.draw_mpl(circuit, decimals=None, fig=plt.figure(10))(params)
```

**Description of the Change:**

* We have chosen to pass the matplotlib figure and not the matplotlib axes as figure properties are used in the current implementation. See for example `MPLDrawer._text_dims`.
* Since `draw_mpl` sets the figure size explicitely we also do this when using an existing figure
* Use `fig` as a keyword argument only

@josh146 

**Benefits:**

Re-using figure windows prevents many open windows in interactive environments such as Spyder

**Possible Drawbacks:**

PR is fully backwards compatible.

**Related GitHub Issues:**

#4950

### CODE REVIEW

**Comment by josh146** [comment]
Thanks @eendebakpt! I will organize a code review for this PR. Out of curiosity, how come you decided to pass `fig` as a keyword argument, as opposed to `ax`? My understanding was that passing axes would provide more flexibility/is more common

**Comment by eendebakpt** [comment]
> Thanks @eendebakpt! I will organize a code review for this PR. Out of curiosity, how come you decided to pass `fig` as a keyword argument, as opposed to `ax`? My understanding was that passing axes would provide more flexibility/is more common

Passing `ax` is indeed more flexible. The reason for passsing `fig` is that the `MPLDrawer` requires a matplotlib figure (the `.fig` property). It is used to set the figure size (here https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/drawer/mpldrawer.py#L272) and some text properties
(here https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/drawer/mpldrawer.py#L564).

I am not familiar with the design and history of the drawing functionality, but perhaps with a little refactoring passing the `ax` instead is also possible.

**Review by albi3ro — COMMENTED** [review]
Thanks for opening this PR. rThis seems like it could be quite a nice feature 🎉 

As you pointed out in the PR description, we do use the figure to autosize text. I do see a couple of options that would allow us to accept the axes object. We could accept both the figure *and* the axes object. Then we could either:

1. Error if axes are provided without a figure.
2. Turn off text auto-sizing and warn the user if axes are provided without a figure

Minor housekeeping tasks:

- [ ]  Add a [changelog](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md) entry and add your name to the list of contributors
- [ ] Add a short test to [test_draw_mpl.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/drawer/test_draw_mpl.py) 
- [ ] Add a short test to [test_tape_mpl.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/drawer/test_tape_mpl.py)
- [ ] Add a test to [test_mpldrawer.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/drawer/test_mpldrawer.py)
- [ ] Add `fix` to the list of keyword arguments for `MPLDrawer`: https://github.com/PennyLaneAI/pennylane/blob/f496a638f74a87a8b941e4b9cdbfaa55b2e6ea5c/pennylane/drawer/mpldrawer.py#L71

Tests can simply be checking the outputted figure is the same object as the inputted figure.

**Comment by eendebakpt** [comment]
> Thanks for opening this PR. rThis seems like it could be quite a nice feature 🎉
> 
> As you pointed out in the PR description, we do use the figure to autosize text. I do see a couple of options that would allow us to accept the axes object. We could accept both the figure _and_ the axes object. Then we could either:
> 
> 1. Error if axes are provided without a figure.
> 2. Turn off text auto-sizing and warn the user if axes are provided without a figure.
> 
> The autosizing is just to make the text look better if a gate has a particularily long label. Most of the time, it won't kick in and make a difference.
> 
> Minor housekeeping tasks:
> 
> * [ ]   Add a [changelog](https://github.com/PennyLaneAI/pennylane/blob/master/doc/releases/changelog-dev.md) entry and add your name to the list of contributors
> * [ ]  Add a short test to [test_draw_mpl.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/drawer/test_draw_mpl.py)
> * [ ]  Add a short test to [test_tape_mpl.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/drawer/test_tape_mpl.py)
> * [ ]  Add a test to [test_mpldrawer.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/drawer/test_mpldrawer.py)
> * [ ]  Add `fix` to the list of keyword arguments for `MPLDrawer`: https://github.com/PennyLaneAI/pennylane/blob/f496a638f74a87a8b941e4b9cdbfaa55b2e6ea5c/pennylane/drawer/mpldrawer.py#L71
> 
> Tests can simply be checking the outputted figure is the same object as the inputted figure.

Housekeeping tasks have been done. I decided to keep the `ax` argument out for now. If there is a use case, we could add it (with either one of your suggestions of handling the resizing)

File: pennylane/drawer/mpldrawer.py [hunk 1]
```diff
@@ -71,6 +71,7 @@ class MPLDrawer:
         wire_options=None (dict): matplotlib configuration options for drawing the wire lines
         figsize=None (Iterable): Allows users to specify the size of the figure manually. Defaults
             to scale with the size of the circuit via ``n_layers`` and ``n_wires``.
+        figsize=None (matplotlib Figure): Allows users to specify the figure window to plot to.
```
Comment (albi3ro) [inline pennylane/drawer/mpldrawer.py:74]:
```suggestion
        fig=None (matplotlib Figure): Allows users to specify the figure window to plot to.
```

**Review by albi3ro — APPROVED** [review]
Looks great. Thanks for making those changes 👍

**Review by lillian542 — APPROVED** [review]
Looks good @eendebakpt! This will go into the 0.34 release next week, which is already partially underway, so it will be merged into a different branch instead of master. I'll change the target branch and fix any conflicts before merging. Thanks for working on this! 🚀

**Pull request merged**